### ISSUE TITLE
Algo_Registry doesn't really work when static linking.

### ISSUE THREAD

**Issue by cdesjardins** [randombit/botan#52]
I don't know if you already know this or not, or if you have a plan to make this work better, so I thought I would just let you know what is happening. Last night I decided to pull the latest from botan, and when I got through all the compile errors, it started running and I started seeing that a bunch of get_*_function were returning NULL. This was because the REGISTER_* macros were not being invoked for anything that I did not specifically refer to in my application. When static linking to a library the linker only pulls in files that are "required", so things that I do not specifically refer to do not get linked into my application, and as such the REGISTER_ macros do not run for those classes. To work around it I just declared an instance of anything that I was going to lookup. In the end it was getting a bit out of control so I decided to revert back to the commit I was pointing to previously, which still uses the old algorithm_factory.

**Comment by randombit** [comment]
I did not know about this, thanks for opening the issue. I suppose I had just assumed that the linker would include any objects from static archives that contained globals, considering especially there are intended be side effects of the constructors. Blargh. It appears this limitation also applies to functions marked with the gcc `constructor` attribute, so no out there either.

In the short term using `-Wl,--whole-archive -lbotan -Wl,--no-whole-archive` ... at link time will work and more or less get you the same effect as previously (as the global state pulled in virtually all objects by pulling in the lookup tables in the engine code which reference constructors defined in those objects).

Perhaps the solution is to always compile the static library via the amalgamation? That way there is only a single object to be referenced so all the registration objects will be pulled in. Somewhat unfortunate on the resulting binary sizes though.

**Comment by cdesjardins** [comment]
Thanks for getting back to me, I didn't know about the amalgamation, it looks like it should work for now (haven't tried it yet of course). Luckily I am not too concerned about binary size at this time, but right now things are still very early stages. 

I might also suggest maybe consider bringing back the LibraryInit class you had, I know you are trying to get rid of it, but maybe it could be used to resolve this issue as well, or maybe just make a StaticLibraryInit class or something like that? Or just tell static linkers to use the amalgamation :D

**Comment by cdesjardins** [comment]
Just an update on this, I was able to get the amalgamation build to work pretty easily on linux, but on windows it required some hacks, most notable is:
```
\boost_1_57_0\boost/asio/detail/socket_types.hpp(24) : fatal error C1189: #error :  WinSock.h has already been included
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\BIN\cl.EXE"' : return code '0x2'
Stop.
```

To work around this error I just moved the boost includes to the very top of botan_all.cpp just under the include of botan_all.h

I also had to disable a few extra modules that didn't compile:
```
configure.py --disable-shared --disable-modules=mce,selftest,tls,ffi,mceies,curve25519 --prefix=botan\install --enable-debug --cpu=i386 --via-amalgamation --with-boost
```

**Comment by randombit** [comment]
Still not resolved, for instance #58

**Comment by webmaster128** [comment]
I think this issue is covered and more precise in the follow-up discussion #128. If there are bugs left when amalgamation is used, please open a new specific issue.
